kotlinx-coroutines-core/common/src/internal/OnDemandAllocatingPool.kt

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2016-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+import kotlinx.atomicfu.*
+
+/**
+ * A thread-safe resource pool.
+ *
+ * [maxCapacity] is the maximum amount of elements.
+ * [create] is the function that creates a new element.
+ */
+internal class OnDemandAllocatingPool<T>(
+    private val maxCapacity: Int,
+    private val create: (Int) -> T
+) {
+    /** Number of existing elements + isClosed flag in the highest bit.
+     * Once the flag is set, the value is guaranteed not to change anymore. */
```

Comment:
I would stick with the default Java convention, that autoformats this comment into the following form:
```
/*
 * Comment
 */
```
I've stressed our (Kotlin) auto-formatter for a while, and it seems like it just doesn't work at all, I'll file all the bugs separately